cmd/gateway/s3/gateway-s3-sse.go

```diff
@@ -0,0 +1,629 @@
+				continue
+			}
+			prefixes = append(prefixes, p)
+		}
+		if (len(objects) > maxKeys) || !loi.IsTruncated {
+			break
+		}
+	}
+
+	loi.IsTruncated = isTruncated
+	loi.ContinuationToken = continuationToken
+	loi.Objects = make([]minio.ObjectInfo, len(objects))
+	loi.Prefixes = make([]string, 0)
+	for _, obj := range objects {
+		loi.NextContinuationToken = obj.Name
+		loi.Objects = append(loi.Objects, obj)
+	}
+	for _, pfx := range prefixes {
+		if pfx != prefix {
+			loi.Prefixes = append(loi.Prefixes, pfx)
+		}
+	}
+	return loi, nil
+}
+
+// GetObject reads an object from S3. Supports additional
+// parameters like offset and length which are synonymous with
+// HTTP Range requests.
+// In the case of multi-part uploads that were encrypted at the gateway, the objects
+// are stored in a custom format at the backend with each part as an individual object
+// and piped to the writer.
+func (l *s3EncObjects) GetObject(ctx context.Context, bucket string, key string, startOffset int64, length int64, writer io.Writer, etag string, o minio.ObjectOptions) error {
+	// pass through encryption
+	if len(minio.GlobalGatewaySSE) == 0 {
+		return l.s3Objects.GetObject(ctx, bucket, key, startOffset, length, writer, etag, o)
+	}
+
+	dmeta, err := l.getDareMetadata(ctx, bucket, getGWMetaPath(key))
+	if err != nil {
+		return l.s3Objects.GetObject(ctx, bucket, key, startOffset, length, writer, etag, o)
+	}
+	var partStartIndex int
+	var partStartOffset = startOffset
+	// Skip parts until final offset maps to a particular part offset.
+	for i, part := range dmeta.Parts {
+		decryptedSize, err := sio.DecryptedSize(uint64(part.Size))
+		if err != nil {
+			return err
+		}
+		partStartIndex = i
+
+		// Offset is smaller than size we have reached the
+		// proper part offset, break out we start from
+		// this part index.
+		if partStartOffset < int64(decryptedSize) {
+			break
+		}
+		// Continue to look for next part.
+		partStartOffset -= int64(decryptedSize)
+	}
+	startSeqNum := partStartOffset / minio.SSEDAREPackageBlockSize
+	partEncRelOffset := int64(startSeqNum) * (minio.SSEDAREPackageBlockSize + minio.SSEDAREPackageMetaSize)
+
+	var size int64
+
+	// concatenate parts stored as separate objects into writer
+	for i, part := range dmeta.Parts {
+		//skip parts before start offset
+		if i < partStartIndex {
+			continue
+		}
+
+		pInfo, err := l.s3Objects.GetObjectInfo(ctx, bucket, part.Name, o)
+		if err != nil || pInfo.ETag != part.ETag {
+			logger.LogIf(ctx, err)
+			return minio.ObjectNotFound{
+				Bucket: bucket,
+				Object: key,
+			}
+		}
+		partLength := pInfo.Size - partEncRelOffset
+		size += partLength
+		if size > length {
+			partLength -= (size - length)
+		}
+		// Initialize pipe.
+		pipeReader, pipeWriter := io.Pipe()
+
+		var reader io.Reader = pipeReader
+		pInfo.Reader, err = hash.NewReader(reader, partLength, "", "")
+		pInfo.Writer = pipeWriter
+
+		var wg = &sync.WaitGroup{}
+		go func(pInfo minio.ObjectInfo) {
+			wg.Add(1)
```

Comment:
Never call `wg.Add(1)` **inside** goroutine. It's not guaranteed to be called ever.

Reply:
removed this.